flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala

```diff
@@ -411,6 +411,23 @@ class TaskManager(
             log.debug(s"Cannot find task to fail for execution ${executionID})")
           }
 
+        // stops a task
+        case StopTask(executionID) =>
+          val task = runningTasks.get(executionID)
+          if (task != null) {
+            try {
+              task.stopExecution()
```

Comment:
Would it be possible, to change the interface to an abstract class that provides a protected member `isRunning` (or similar) and implements `public FINAL void stop()` method by setting `isRunning = false`. Thus, we would ensure, that the call is non-blocking. The user needs to check `isRunning` flag in `run()` method.